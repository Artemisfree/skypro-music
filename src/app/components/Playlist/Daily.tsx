import React, { useEffect, useRef, useState } from 'react'
import { useDispatch, useSelector } from 'react-redux'
import { RootState } from '@/store/store'
import styles from './Playlist.module.css'
import { Track } from '@/types/types'
import {
	setCurrentTrack,
	playTrack,
	pauseTrack,
    updateCurrentTime,
    setPlayingState,
} from '@/store/features/currentTrackSlice'
import { setTracks } from '@/store/features/playlistSlice'
import { getSelectionById, getTrackById } from '@/app/api'

const DailyPlaylist: React.FC = () => {
	const [loading, setLoading] = useState<boolean>(false)
	const [error, setError] = useState<string | null>(null)
	const dispatch = useDispatch()
	const { currentTrack, isPlaying, currentTime } = useSelector(
		(state: RootState) => state.currentTrack
	)
	const { tracks } = useSelector((state: RootState) => state.playlist)

    const audioRef = useRef<HTMLAudioElement>(null)

	useEffect(() => {
		const fetchDayTracks = async () => {
			setLoading(true)
			try {
				const selection = await getSelectionById(2)
				const trackPromises = selection.data.items.map((id: number) =>
					getTrackById(id)
				)
				const dayTracksResponses = await Promise.all(trackPromises)
				const dayTracks = dayTracksResponses.map(response => response.data)
				dispatch(setTracks(dayTracks))
			} catch (error) {
				setError('Не удалось загрузить треки. Пожалуйста, попробуйте позже.')
			} finally {
				setLoading(false)
			}
		}

		fetchDayTracks()
	}, [dispatch])

	const handleTrackClick = (track: Track) => {
		if (currentTrack?._id === track._id) {
			if (isPlaying) {
				dispatch(pauseTrack())
			} else {
				dispatch(playTrack())
			}
		} else {
			dispatch(setCurrentTrack({ ...track }))
		}
	}

	const formatDuration = (seconds: number) => {
		const minutes = Math.floor(seconds / 60)
		const remainingSeconds = seconds % 60
		return `${minutes}:${remainingSeconds < 10 ? '0' : ''}${remainingSeconds}`
	}

	useEffect(() => {
		if (currentTrack) {
			localStorage.setItem('currentTrack', JSON.stringify(currentTrack))
			localStorage.setItem('currentTime', JSON.stringify(currentTime))
			localStorage.setItem('isPlaying', JSON.stringify(isPlaying))
		}
	}, [currentTrack, currentTime, isPlaying])

	useEffect(() => {
		const savedTrack = localStorage.getItem('currentTrack')
		const savedTime = localStorage.getItem('currentTime')
		const savedIsPlaying = localStorage.getItem('isPlaying')

		if (savedTrack && savedTime && savedIsPlaying !== null) {
			const track = JSON.parse(savedTrack)
			const time = JSON.parse(savedTime)
			const playing = JSON.parse(savedIsPlaying)
			dispatch(setCurrentTrack(track))
			dispatch(updateCurrentTime(time))
			dispatch(setPlayingState(playing))
			if (audioRef.current) {
				audioRef.current.currentTime = time
				if (playing) {
					audioRef.current.play()
				}
			}
		}
	}, [dispatch])

	useEffect(() => {
		const handleTimeUpdate = () => {
			if (audioRef.current) {
				const currentTime = audioRef.current.currentTime
				dispatch(updateCurrentTime(currentTime))
				localStorage.setItem('currentTime', JSON.stringify(currentTime))
			}
		}
		const audioElement = audioRef.current
		if (audioElement) {
			audioElement.addEventListener('timeupdate', handleTimeUpdate)
		}
		return () => {
			if (audioElement) {
				audioElement.removeEventListener('timeupdate', handleTimeUpdate)
			}
		}
	}, [dispatch])

	if (error) {
		return (
			<div>
				<p>{error}</p>
			</div>
		)
	}

	return (
		<div className={styles.centerblock__content}>
			<div className={styles.content__title}>
				<div className={styles.col01}>ТРЕК</div>
				<div className={styles.col02}>ИСПОЛНИТЕЛЬ</div>
				<div className={styles.col03}>АЛЬБОМ</div>
				<div className={styles.col04}>
					<svg className={styles.playlist_title__svg}>
						<use xlinkHref='img/icon/sprite.svg#icon-watch'></use>
					</svg>
				</div>
			</div>
			<div className={styles.content__playlist}>
				{Array.isArray(tracks) &&
					tracks.map((track, index) => (
						<div
							className={`${styles.playlist__track} ${
								currentTrack?._id === track._id ? styles.currentTrack : ''
							}`}
							key={track._id || index}
							onClick={() => handleTrackClick(track)}
						>
							<div className={styles.track__title}>
								<div className={styles.track__title_image}>
									{currentTrack?._id === track._id ? (
										<div
											className={`${
												isPlaying ? styles.pulsingDot : styles.staticDot
											}`}
										></div>
									) : (
										<svg className={styles.track__title_svg}>
											<use xlinkHref='img/icon/sprite.svg#icon-note'></use>
										</svg>
									)}
								</div>
								<div className={styles.track__title_text}>
									<a className={styles.track__title_link} href='#'>
										{track.name}{' '}
										<span className={styles.track__title_span}></span>
									</a>
								</div>
							</div>
							<div className={styles.track__author}>
								<a className={styles.track__author_link} href='#'>
									{track.author}
								</a>
							</div>
							<div className={styles.track__album}>
								<a className={styles.track__album_link} href='#'>
									{track.album}
								</a>
							</div>
							<div className={styles.track__like}>
								<span className={styles.track__time_text}>
									{formatDuration(track.duration_in_seconds)}
								</span>
							</div>
						</div>
					))}
			</div>
		</div>
	)
}

export default DailyPlaylist
