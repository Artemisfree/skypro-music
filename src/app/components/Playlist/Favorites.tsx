import React, { useEffect, useRef, useState } from 'react'
import { useDispatch, useSelector } from 'react-redux'
import { RootState } from '@/store/store'
import styles from './Playlist.module.css'
import { Track } from '@/types/types'
import {
	setCurrentTrack,
	playTrack,
	pauseTrack,
	updateCurrentTime,
	setPlayingState,
} from '@/store/features/currentTrackSlice'
import {
	setTracks,
	removeTrack,
	updateTrackLikeStatus,
} from '@/store/features/playlistSlice'
import { getAllFavoriteTracks, removeTrackFromFavorites } from '@/app/api'

const Favorites: React.FC = () => {
	const [loading, setLoading] = useState<boolean>(false)
	const [error, setError] = useState<string | null>(null)
	const dispatch = useDispatch()
	const { currentTrack, isPlaying, currentTime } = useSelector(
		(state: RootState) => state.currentTrack
	)
	const { tracks } = useSelector((state: RootState) => state.playlist)
	const audioRef = useRef<HTMLAudioElement>(null)

	const getAccessToken = (): string | null =>
		localStorage.getItem('accessToken')
	const getRefreshToken = (): string | null =>
		localStorage.getItem('refreshToken')

	useEffect(() => {
		const fetchFavoriteTracks = async () => {
			const accessToken = getAccessToken()
			const refreshToken = getRefreshToken()
			if (!accessToken || !refreshToken) {
				setError('Пожалуйста, авторизуйтесь для доступа к избранным трекам.')
				return
			}

			try {
				const response = await getAllFavoriteTracks(accessToken, refreshToken)
				dispatch(setTracks(response.data))
			} catch (error) {
				setError(
					'Не удалось загрузить избранные треки. Пожалуйста, попробуйте позже.'
				)
			}
		}

		fetchFavoriteTracks()
	}, [dispatch])

	const handleTrackClick = (track: Track) => {
		if (currentTrack?._id === track._id) {
			if (isPlaying) {
				dispatch(pauseTrack())
			} else {
				dispatch(playTrack())
			}
		} else {
			dispatch(setCurrentTrack({ ...track }))
		}
	}

	const handleDislikeClick = async (e: React.MouseEvent, track: Track) => {
		e.stopPropagation()
		const accessToken = getAccessToken()
		const refreshToken = getRefreshToken()

		if (!accessToken || !refreshToken) {
			setError('Пожалуйста, авторизуйтесь для снятия лайка.')
			return
		}

		const updatedTrack = { ...track, isLiked: false }
		dispatch(updateTrackLikeStatus(updatedTrack))

		if (currentTrack?._id === track._id) {
			dispatch(setCurrentTrack(updatedTrack))
		}

		setLoading(true)
		try {
			await removeTrackFromFavorites(track._id, accessToken, refreshToken)
			dispatch(removeTrack(track._id))
		} catch (err) {
			setError('Не удалось снять лайк. Пожалуйста, попробуйте позже.')
			dispatch(updateTrackLikeStatus(track))
			if (currentTrack?._id === track._id) {
				dispatch(setCurrentTrack(track))
			}
		} finally {
			setLoading(false)
		}
	}

	const formatDuration = (seconds: number) => {
		const minutes = Math.floor(seconds / 60)
		const remainingSeconds = seconds % 60
		return `${minutes}:${remainingSeconds < 10 ? '0' : ''}${remainingSeconds}`
	}

	useEffect(() => {
		if (currentTrack) {
			localStorage.setItem('currentTrack', JSON.stringify(currentTrack))
			localStorage.setItem('currentTime', JSON.stringify(currentTime))
			localStorage.setItem('isPlaying', JSON.stringify(isPlaying))
		}
	}, [currentTrack, currentTime, isPlaying])

	useEffect(() => {
		const savedTrack = localStorage.getItem('currentTrack')
		const savedTime = localStorage.getItem('currentTime')
		const savedIsPlaying = localStorage.getItem('isPlaying')

		if (savedTrack && savedTime && savedIsPlaying !== null) {
			const track = JSON.parse(savedTrack)
			const time = JSON.parse(savedTime)
			const playing = JSON.parse(savedIsPlaying)
			dispatch(setCurrentTrack(track))
			dispatch(updateCurrentTime(time))
			dispatch(setPlayingState(playing))
			if (audioRef.current) {
				audioRef.current.currentTime = time
				if (playing) {
					audioRef.current.play().catch(error => {
						console.error('Error playing audio:', error)
					})
				}
			}
		}
	}, [dispatch])

	useEffect(() => {
		const handleTimeUpdate = () => {
			if (audioRef.current) {
				const currentTime = audioRef.current.currentTime
				dispatch(updateCurrentTime(currentTime))
				localStorage.setItem('currentTime', JSON.stringify(currentTime))
			}
		}
		const audioElement = audioRef.current
		if (audioElement) {
			audioElement.addEventListener('timeupdate', handleTimeUpdate)
		}
		return () => {
			if (audioElement) {
				audioElement.removeEventListener('timeupdate', handleTimeUpdate)
			}
		}
	}, [dispatch])

	if (error) {
		return (
			<div>
				<p>{error}</p>
			</div>
		)
	}

	return (
		<div className={styles.centerblock__content}>
			<div className={styles.content__title}>
				<div className={styles.col01}>ТРЕК</div>
				<div className={styles.col02}>ИСПОЛНИТЕЛЬ</div>
				<div className={styles.col03}>АЛЬБОМ</div>
				<div className={styles.col04}>
					<svg className={styles.playlist_title__svg}>
						<use xlinkHref='img/icon/sprite.svg#icon-watch'></use>
					</svg>
				</div>
			</div>
			<div className={styles.content__playlist}>
				{Array.isArray(tracks) && tracks.length > 0 ? (
					tracks.map(track => (
						<div
							className={`${styles.playlist__track} ${
								currentTrack?._id === track._id ? styles.currentTrack : ''
							}`}
							key={track._id}
							onClick={() => handleTrackClick(track)}
						>
							<div className={styles.track__title}>
								<div className={styles.track__title_image}>
									{currentTrack?._id === track._id ? (
										<div
											className={`${
												isPlaying ? styles.pulsingDot : styles.staticDot
											}`}
										></div>
									) : (
										<svg className={styles.track__title_svg}>
											<use xlinkHref='img/icon/sprite.svg#icon-note'></use>
										</svg>
									)}
								</div>
								<div className={styles.track__title_text}>
									<a className={styles.track__title_link} href='#'>
										{track.name}{' '}
										<span className={styles.track__title_span}></span>
									</a>
								</div>
							</div>
							<div className={styles.track__author}>
								<a className={styles.track__author_link} href='#'>
									{track.author}
								</a>
							</div>
							<div className={styles.track__album}>
								<a className={styles.track__album_link} href='#'>
									{track.album}
								</a>
							</div>
							<div
								className={`${styles.track__like} _btn-icon`}
								onClick={e => handleDislikeClick(e, track)}
							>
								<svg className={styles.track__time_svg}>
									<use xlinkHref='img/icon/sprite.svg#icon-liked'></use>
								</svg>
								<span className={styles.track__time_text}>
									{formatDuration(track.duration_in_seconds)}
								</span>
							</div>
						</div>
					))
				) : (
					<div className={styles.noTracksFound}>Треки не найдены</div>
				)}
			</div>
		</div>
	)
}

export default Favorites
