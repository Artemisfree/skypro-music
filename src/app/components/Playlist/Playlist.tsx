import React from 'react'
import styles from './Playlist.module.css'

const Playlist = () => {
	return (
		<div className={styles.centerblock__content}>
			<div className={styles.content__title}>
				<div className={styles.col01}>Трек</div>
				<div className={styles.col02}>Исполнитель</div>
				<div className={styles.col03}>Альбом</div>
				<div className={styles.col04}>
					<svg className={styles.playlist_title__svg}>
						<use xlinkHref='img/icon/sprite.svg#icon-watch'></use>
					</svg>
				</div>
			</div>
			<div className={styles.content__playlist}>
				<div className={styles.playlist__item}>
					<div className={styles.playlist__track}>
						<div className={styles.track__title}>
							<div className={styles.track__title_image}>
								<svg className={styles.track__title_svg}>
									<use xlinkHref='img/icon/sprite.svg#icon-note'></use>
								</svg>
							</div>
							<div className={styles.track__title_text}>
								<a className={styles.track__title_link} href='http://'>
									Guilt <span className={styles.track__title_span}></span>
								</a>
							</div>
						</div>
						<div className={styles.track__author}>
							<a className={styles.track__author_link} href='http://'>
								Nero
							</a>
						</div>
						<div className={styles.track__album}>
							<a className={styles.track__album_link} href='http://'>
								Welcome Reality
							</a>
						</div>
						<div className={styles.track__time}>
							<svg className={styles.track__time_svg}>
								<use xlinkHref='img/icon/sprite.svg#icon-like'></use>
							</svg>
							<span className={styles.track__time_text}>4:44</span>
						</div>
					</div>
					<div className={styles.playlist__track}>
						<div className={styles.track__title}>
							<div className={styles.track__title_image}>
								<svg className={styles.track__title_svg}>
									<use xlinkHref='img/icon/sprite.svg#icon-note'></use>
								</svg>
							</div>
							<div className={styles.track__title_text}>
								<a className={styles.track__title_link} href='http://'>
									Elektro <span className={styles.track__title_span}></span>
								</a>
							</div>
						</div>
						<div className={styles.track__author}>
							<a className={styles.track__author_link} href='http://'>
								Dynoro, Outwork, Mr. Gee
							</a>
						</div>
						<div className={styles.track__album}>
							<a className={styles.track__album_link} href='http://'>
								Elektro
							</a>
						</div>
						<div className={styles.track__time}>
							<svg className={styles.track__time_svg}>
								<use xlinkHref='img/icon/sprite.svg#icon-like'></use>
							</svg>
							<span className={styles.track__time_text}>2:22</span>
						</div>
					</div>
					<div className={styles.playlist__track}>
						<div className={styles.track__title}>
							<div className={styles.track__title_image}>
								<svg className={styles.track__title_svg}>
									<use xlinkHref='img/icon/sprite.svg#icon-note'></use>
								</svg>
							</div>
							<div className={styles.track__title_text}>
								<a className={styles.track__title_link} href='http://'>
									I’m Fire <span className={styles.track__title_span}></span>
								</a>
							</div>
						</div>
						<div className={styles.track__author}>
							<a className={styles.track__author_link} href='http://'>
								Ali Bakgor
							</a>
						</div>
						<div className={styles.track__album}>
							<a className={styles.track__album_link} href='http://'>
								I’m Fire
							</a>
						</div>
						<div className={styles.track__time}>
							<svg className={styles.track__time_svg}>
								<use xlinkHref='img/icon/sprite.svg#icon-like'></use>
							</svg>
							<span className={styles.track__time_text}>2:22</span>
						</div>
					</div>
					<div className={styles.playlist__track}>
						<div className={styles.track__title}>
							<div className={styles.track__title_image}>
								<svg className={styles.track__title_svg}>
									<use xlinkHref='img/icon/sprite.svg#icon-note'></use>
								</svg>
							</div>
							<div className={styles.track__title_text}>
								<a className={styles.track__title_link} href='http://'>
									Non Stop <span className={styles.track__title_span}></span>
								</a>
							</div>
						</div>
						<div className={styles.track__author}>
							<a className={styles.track__author_link} href='http://'>
								Стоункат, Psychopath
							</a>
						</div>
						<div className={styles.track__album}>
							<a className={styles.track__album_link} href='http://'>
								Non Stop
							</a>
						</div>
						<div className={styles.track__time}>
							<svg className={styles.track__time_svg}>
								<use xlinkHref='img/icon/sprite.svg#icon-like'></use>
							</svg>
							<span className={styles.track__time_text}>4:12</span>
						</div>
					</div>
				</div>
			</div>
		</div>
	)
}

export default Playlist
