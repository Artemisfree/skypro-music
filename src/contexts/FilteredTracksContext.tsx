'use client'

import React, { createContext, useState, useContext } from 'react'

interface FilteredTracksContextType {
    activeGenres: string[]
    setActiveGenres: React.Dispatch<React.SetStateAction<string[]>>
    activeAuthors: string[]
    setActiveAuthors: React.Dispatch<React.SetStateAction<string[]>>
    searchKeyword: string
    setSearchKeyword: React.Dispatch<React.SetStateAction<string>>
    sortOrder: string
    setSortOrder: React.Dispatch<React.SetStateAction<string>>
}

const FilteredTracksContext = createContext<FilteredTracksContextType | undefined>(undefined)

export const FilteredTracksProvider: React.FC<{ children: React.ReactNode }> = ({ children }) => {
    const [activeGenres, setActiveGenres] = useState<string[]>([])
    const [activeAuthors, setActiveAuthors] = useState<string[]>([])
    const [searchKeyword, setSearchKeyword] = useState<string>('')
    const [sortOrder, setSortOrder] = useState<string>('default')

    return (
        <FilteredTracksContext.Provider value={{ activeGenres, setActiveGenres, activeAuthors, setActiveAuthors, searchKeyword, setSearchKeyword, sortOrder, setSortOrder }}>
            {children}
        </FilteredTracksContext.Provider>
    )
}

export const useFilteredTracks = (): FilteredTracksContextType => {
    const context = useContext(FilteredTracksContext)
    if (!context) {
        throw new Error('useFilteredTracks must be used within a FilteredTracksProvider')
    }
    return context
}
