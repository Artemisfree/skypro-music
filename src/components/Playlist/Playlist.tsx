'use client'

import React, { useEffect, useMemo, useRef, useState } from 'react'
import {
	addTrackToFavorites,
	removeTrackFromFavorites,
	getAllTracks,
	getSelectionById,
	getTrackById,
} from '../../../api/api'
import styles from './Playlist.module.css'
import { useDispatch, useSelector } from 'react-redux'
import { RootState } from '@/store/store'
import {
	setCurrentTrack,
	updateCurrentTime,
} from '@/store/features/currentTrackSlice'
import {
	setTracks,
	updateTrackLikeStatus,
} from '@/store/features/playlistSlice'
import { Track } from '@/types/types'
import { useFilteredTracks } from '@/contexts/FilteredTracksContext'

type PlaylistProps = {
	playlistId?: number
	tracks: Track[]
}

const Modal: React.FC<{ message: string; onClose: () => void }> = ({
	message,
	onClose,
}) => (
	<div className={styles.modalOverlay}>
		<div className={styles.modal}>
			<p>{message}</p>
			<button onClick={onClose}>Закрыть</button>
		</div>
	</div>
)

const Playlist: React.FC<PlaylistProps> = ({ playlistId, tracks }) => {
	const [error, setError] = useState<string | null>(null)
	const [isModalVisible, setModalVisible] = useState<boolean>(false)
	const [likeError, setLikeError] = useState<string | null>(null)
	const [likedTracks, setLikedTracks] = useState<{ [key: number]: boolean }>({})
	const dispatch = useDispatch()
	const { currentTrack, isPlaying } = useSelector(
		(state: RootState) => state.currentTrack
	)
	const playlistTracks = useSelector(
		(state: RootState) => state.playlist.tracks
	)

	useEffect(() => {
		dispatch(setTracks(tracks))
	}, [dispatch, tracks])

	const { activeGenres, activeAuthors, searchKeyword, sortOrder } =
		useFilteredTracks()
	const audioRef = useRef<HTMLAudioElement>(null)

	const handleTrackClick = (track: Track) => {
		if (currentTrack?._id !== track._id) {
			dispatch(setCurrentTrack({ ...track }))
		}
	}

	const saveLikedState = (trackId: number, isLiked: boolean) => {
		const likedTracks = JSON.parse(localStorage.getItem('likedTracks') || '{}')
		likedTracks[trackId] = isLiked
		localStorage.setItem('likedTracks', JSON.stringify(likedTracks))
		setLikedTracks(likedTracks)
	}

	const getLikedState = (trackId: number) => {
		const likedTracks = JSON.parse(localStorage.getItem('likedTracks') || '{}')
		return likedTracks[trackId] || false
	}

	const handleLikeClick = async (track: Track) => {
		const accessToken = localStorage.getItem('accessToken')
		const refreshToken = localStorage.getItem('refreshToken')

		if (!accessToken || !refreshToken) {
			setModalVisible(true)
			return
		}

		try {
			if (likedTracks[track._id]) {
				await removeTrackFromFavorites(track._id)
			} else {
				await addTrackToFavorites(track._id)
			}
			const updatedTracks = tracks.map(t =>
				t._id === track._id ? { ...t, isLiked: !t.isLiked } : t
			)
			dispatch(setTracks(updatedTracks))
			saveLikedState(track._id, !likedTracks[track._id])
			dispatch(
				updateTrackLikeStatus({
					_id: track._id,
					isLiked: !likedTracks[track._id],
				})
			)
		} catch (error) {
			setLikeError('Не удалось обновить лайк. Пожалуйста, попробуйте позже.')
			console.error('Ошибка при обновлении лайка:', error)
		}
	}

	useEffect(() => {
		const fetchTracks = async () => {
			try {
				let allTracks: Track[] = []
				if (playlistId) {
					const selection = await getSelectionById(playlistId)
					const trackPromises = selection.data.items.map((id: number) =>
						getTrackById(id)
					)
					const playlistTracksResponses = await Promise.all(trackPromises)
					allTracks = playlistTracksResponses
						.map(response => response.data)
						.map((track: Track) => ({
							...track,
							isLiked: getLikedState(track._id),
						}))
				} else {
					const response = await getAllTracks()
					allTracks = response.data.map((track: Track) => ({
						...track,
						isLiked: getLikedState(track._id),
					}))
				}
				dispatch(setTracks(allTracks))
				const initialLikedTracks = allTracks.reduce(
					(acc: { [key: number]: boolean }, track: Track) => {
						acc[track._id] = getLikedState(track._id)
						return acc
					},
					{}
				)
				setLikedTracks(initialLikedTracks)
			} catch (error) {
				setError('Не удалось загрузить треки. Пожалуйста, попробуйте позже.')
			}
		}
		fetchTracks()
	}, [dispatch, playlistId])

	useEffect(() => {
	}, [playlistTracks])

	const filteredTracks = useMemo(() => {
		return Array.isArray(playlistTracks)
			? playlistTracks.filter(track => {
					const matchesGenre =
						activeGenres.length === 0 ||
						track.genre.some(genre => activeGenres.includes(genre))
					const matchesAuthor =
						activeAuthors.length === 0 || activeAuthors.includes(track.author)
					const matchesSearch =
						!searchKeyword ||
						track.name.toLowerCase().includes(searchKeyword.toLowerCase()) ||
						track.author.toLowerCase().includes(searchKeyword.toLowerCase())
					return matchesGenre && matchesAuthor && matchesSearch
			  })
			: []
	}, [playlistTracks, activeGenres, activeAuthors, searchKeyword])

	const sortedTracks = useMemo(() => {
		return filteredTracks.sort((a, b) => {
			if (sortOrder === 'newest') {
				return (
					new Date(b.release_date).getTime() -
					new Date(a.release_date).getTime()
				)
			}
			if (sortOrder === 'oldest') {
				return (
					new Date(a.release_date).getTime() -
					new Date(b.release_date).getTime()
				)
			}
			return 0
		})
	}, [filteredTracks, sortOrder])

	const formatDuration = (seconds: number) => {
		const minutes = Math.floor(seconds / 60)
		const remainingSeconds = seconds % 60
		return `${minutes}:${remainingSeconds < 10 ? '0' : ''}${remainingSeconds}`
	}

	const closeModal = () => {
		setModalVisible(false)
		setError(null)
		setLikeError(null)
	}

	useEffect(() => {
		const handleTimeUpdate = () => {
			if (audioRef.current) {
				const currentTime = audioRef.current.currentTime
				dispatch(updateCurrentTime(currentTime))
			}
		}
		const audioElement = audioRef.current
		if (audioElement) {
			audioElement.addEventListener('timeupdate', handleTimeUpdate)
		}
		return () => {
			if (audioElement) {
				audioElement.removeEventListener('timeupdate', handleTimeUpdate)
			}
		}
	}, [dispatch])

	if (error) {
		return (
			<div>
				<Modal message={error} onClose={closeModal} />
			</div>
		)
	}

	return (
		<div className={styles.centerblock__content}>
			{isModalVisible && (
				<Modal
					message='Пожалуйста, авторизуйтесь для того, чтобы ставить лайки.'
					onClose={closeModal}
				/>
			)}
			{likeError && <Modal message={likeError} onClose={closeModal} />}
			<div className={styles.content__title}>
				<div className={styles.col01}>ТРЕК</div>
				<div className={styles.col02}>ИСПОЛНИТЕЛЬ</div>
				<div className={styles.col03}>АЛЬБОМ</div>
				<div className={styles.col04}>
					<svg className={styles.playlist_title__svg}>
						<use href='/img/icon/sprite.svg#icon-watch'></use>
					</svg>
				</div>
			</div>
			<div className={styles.content__playlist}>
				{Array.isArray(sortedTracks) && sortedTracks.length > 0 ? (
					sortedTracks.map((track: Track) => (
						<div
							className={`${styles.playlist__track} ${
								currentTrack?._id === track._id ? styles.currentTrack : ''
							}`}
							key={track._id}
							onClick={() => handleTrackClick(track)}
						>
							<div className={styles.track__title}>
								<div className={styles.track__title_image}>
									{currentTrack?._id === track._id ? (
										<div className={styles.staticDot}></div>
									) : (
										<svg className={styles.track__title_svg}>
											<use href='/img/icon/sprite.svg#icon-note'></use>
										</svg>
									)}
								</div>
								<div className={styles.track__title_text}>
									{track.name}{' '}
									<span className={styles.track__title_span}></span>
								</div>
							</div>
							<div className={styles.track__author}>{track.author}</div>
							<div className={styles.track__album}>{track.album}</div>
							<div
								className={`${styles.track__like} _btn-icon`}
								onClick={e => {
									e.stopPropagation()
									handleLikeClick(track)
								}}
							>
								{localStorage.getItem('accessToken') ? (
									<>
										<svg className={styles.track__time_svg}>
											<use
												href={
													likedTracks[track._id]
														? '/img/icon/sprite.svg#icon-liked'
														: '/img/icon/sprite.svg#icon-like'
												}
											></use>
										</svg>
										<span className={styles.track__time_text}>
											{formatDuration(track.duration_in_seconds)}
										</span>
									</>
								) : (
									<svg className={styles.track__time_svg}>
										<use href='/img/icon/sprite.svg#icon-watch'></use>
									</svg>
								)}
							</div>
						</div>
					))
				) : (
					<div className={styles.noTracksFound}>Треки не найдены</div>
				)}
			</div>
		</div>
	)
}

export default Playlist
