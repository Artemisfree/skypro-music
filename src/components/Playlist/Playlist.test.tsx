import React from 'react'
import renderer from 'react-test-renderer'
import { Provider } from 'react-redux'
import { configureStore } from '@reduxjs/toolkit'
import currentTrackReducer from '@/store/features/currentTrackSlice'
import playlistReducer from '@/store/features/playlistSlice'
import authReducer from '@/store/features/authSlice'
import { FilteredTracksProvider } from '@/contexts/FilteredTracksContext'
import Playlist from './Playlist'

jest.mock('@/app/api', () => ({
	getAllTracks: jest.fn().mockResolvedValue({
		data: [
			{
				_id: '1',
				name: 'Track1',
				author: 'Author1',
				genre: ['Genre1'],
				album: 'Album1',
				release_date: '2020-01-01',
				duration_in_seconds: 180,
				isLiked: false,
			},
			{
				_id: '2',
				name: 'Track2',
				author: 'Author2',
				genre: ['Genre2'],
				album: 'Album2',
				release_date: '2021-01-01',
				duration_in_seconds: 210,
				isLiked: true,
			},
		],
	}),
	addTrackToFavorites: jest.fn().mockResolvedValue({}),
	removeTrackFromFavorites: jest.fn().mockResolvedValue({}),
	getAllFavoriteTracks: jest.fn().mockResolvedValue({
		data: [
			{
				_id: '2',
				name: 'Track2',
				author: 'Author2',
				genre: ['Genre2'],
				album: 'Album2',
				release_date: '2021-01-01',
				duration_in_seconds: 210,
				isLiked: true,
			},
		],
	}),
	getSelectionById: jest.fn().mockResolvedValue({
		data: {
			items: [1, 2],
		},
	}),
	getTrackById: jest.fn().mockResolvedValue(id => ({
		data: {
			_id: String(id),
			name: `Track${id}`,
			author: `Author${id}`,
			genre: [`Genre${id}`],
			album: `Album${id}`,
			release_date: `2020-01-01`,
			duration_in_seconds: 180,
			isLiked: false,
		},
	})),
}))

const store = configureStore({
	reducer: {
		currentTrack: currentTrackReducer,
		playlist: playlistReducer,
		auth: authReducer,
	},
})

describe('Playlist component', () => {
	it('renders correctly and matches snapshot', async () => {
		const component = renderer.create(
			<Provider store={store}>
				<FilteredTracksProvider>
					<Playlist />
				</FilteredTracksProvider>
			</Provider>
		)

		await renderer.act(async () => {
			await Promise.resolve()
		})

		let tree = component.toJSON()
		expect(tree).toMatchSnapshot()
	})
})
