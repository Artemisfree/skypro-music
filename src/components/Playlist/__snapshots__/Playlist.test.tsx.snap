// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Playlist component renders correctly and matches snapshot 1`] = `
<div
  className="centerblock__content"
>
  <div
    className="content__title"
  >
    <div
      className="col01"
    >
      ТРЕК
    </div>
    <div
      className="col02"
    >
      ИСПОЛНИТЕЛЬ
    </div>
    <div
      className="col03"
    >
      АЛЬБОМ
    </div>
    <div
      className="col04"
    >
      <svg
        className="playlist_title__svg"
      >
        <use
          xlinkHref="img/icon/sprite.svg#icon-watch"
        />
      </svg>
    </div>
  </div>
  <div
    className="content__playlist"
  >
    <div
      className="noTracksFound"
    >
      Треки не найдены
    </div>
  </div>
</div>
`;
